src/pip/_internal/utils/entrypoints.py

```diff
@@ -25,3 +41,34 @@ def _wrapper(args: Optional[List[str]] = None) -> int:
         "running pip directly.\n"
     )
     return main(args)
+
+
+def get_best_invocation_for_this_pip() -> str:
+    """Try to figure out the best way to invoke pip in the current environment."""
+    binary_directory = "Scripts" if WINDOWS else "bin"
+    binary_prefix = os.path.join(sys.prefix, binary_directory)
+
+    # Try to use pip[X[.Y]] names, if those executables for this environment are
+    # the first on PATH with that name.
```

Comment:
These will reintroduce the error where a recommended `pip install -U pip` fails because pip can't update the in-use executable wrapper on Windows. Do we want to do that (and presumably respond "tough, you should have realised" when people encounter it)?

Reply:
Yes, and I've fixed that by updating the upgrade prompt logic to special case handling on Windows.

Reply:
Ooooh! Good catch.

FWIW, we shouldn't bork the pip installation regardless -- since the protection logic is in place. But, yes, we should print a message that would work for sure. :)